File: app/src/main/java/net/squanchy/tweets/service/TwitterService.java
package net.squanchy.tweets.service;

import java.util.List;

import net.squanchy.tweets.domain.view.TweetViewModel;

import io.reactivex.Observable;

import static net.squanchy.support.lang.Lists.filter;
import static net.squanchy.support.lang.Lists.map;

public class TwitterService {

    private final TwitterRepository repository;
    private final TweetModelConverter modelConverter;

    public TwitterService(TwitterRepository repository, TweetModelConverter modelConverter) {
        this.repository = repository;
        this.modelConverter = modelConverter;
    }

    public Observable<List<TweetViewModel>> refresh(String query) {
        return repository.load(query)
                .map(search -> search.tweets)
                .map(list -> filter(list, tweet -> tweet.retweetedStatus == null))
                .map(tweets -> map(tweets, modelConverter::toViewModel));
    }
}

File: app/src/main/java/net/squanchy/tweets/service/TwitterService.kt
package net.squanchy.tweets.service

import io.reactivex.Observable
import io.reactivex.functions.BiFunction
import net.squanchy.service.firestore.FirestoreDbService
import net.squanchy.tweets.domain.view.TwitterScreenViewModel
import net.squanchy.tweets.view.TweetUrlSpanFactory

internal class TwitterService(
    dbService: FirestoreDbService,
    private val factory: TweetUrlSpanFactory
) {

    private val tweets = dbService.twitterView()
        .map { tweets -> tweets.sortedByDescending { it.createdAt } }
        .map { tweets -> tweets.map { mapToViewModel(factory, it) } }

    private val hashtag = dbService.conferenceInfo()
        .map { it.socialHashtag }

    fun refresh(): Observable<TwitterScreenViewModel> {
        return Observable.combineLatest(tweets, hashtag, BiFunction { tweets, hashtag -> TwitterScreenViewModel(hashtag, tweets) })
    }
}
